File: src/main/java/com/strikingly/data/logappenderjava/SqsAppender.java
package com.strikingly.data.logappenderjava;

import com.amazonaws.services.sqs.AmazonSQS;
import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
import com.amazonaws.services.sqs.model.GetQueueUrlResult;
import com.amazonaws.services.sqs.model.SendMessageRequest;
import org.apache.logging.log4j.core.Filter;
import org.apache.logging.log4j.core.Layout;
import org.apache.logging.log4j.core.LogEvent;
import org.apache.logging.log4j.core.appender.AbstractAppender;
import org.apache.logging.log4j.core.config.plugins.Plugin;
import org.apache.logging.log4j.core.config.plugins.PluginAttribute;
import org.apache.logging.log4j.core.config.plugins.PluginElement;
import org.apache.logging.log4j.core.config.plugins.PluginFactory;
import org.apache.logging.log4j.core.layout.PatternLayout;

import java.io.Serializable;

@Plugin(name = "SqsAppender", category = "Core", elementType = "appender", printObject = true)
public class SqsAppender extends AbstractAppender {
    private static AmazonSQS sqsClient;
    private static String sqsUrl = null;

    /* 构造函数 */
    public SqsAppender(String name, Filter filter, Layout<? extends Serializable> layout, boolean ignoreExceptions) {
        super(name, filter, layout, ignoreExceptions);
    }

    @Override
    public void append(LogEvent event) {
        sqsClient.sendMessage(new SendMessageRequest(sqsUrl, getLayout().toSerializable(event).toString()));
    }

    /*  接收配置文件中的参数 */
    @PluginFactory
    public static SqsAppender createAppender(@PluginAttribute("name") String name,
                                             @PluginAttribute("QueueName") String queueName,
                                             @PluginElement("Filter") final Filter filter,
                                             @PluginElement("Layout") Layout<? extends Serializable> layout,
                                             @PluginAttribute("ignoreExceptions") boolean ignoreExceptions) {
        if (name == null) {
            LOGGER.error("no name defined in conf.");
            return null;
        }
        if (layout == null) {
            layout = PatternLayout.createDefaultLayout();
        }
        try {
            String region = System.getProperty("AWS_REGION");
            if (region == null) { region = "cn-north-1"; }

            sqsClient = AmazonSQSClientBuilder.standard().withRegion(region).build();
            GetQueueUrlResult queueUrlResult = sqsClient.getQueueUrl(queueName);
            sqsUrl = queueUrlResult.getQueueUrl();
        } catch (Exception exp) {
            LOGGER.error("SQS name error", exp);
        }

        return new SqsAppender(name, filter, layout, ignoreExceptions);
    }

    public static void main(String[] args) {
        // do nothing
    }
}

File: src/main/scala/com/strikingly/data/logappender/SqsAppender.scala
package com.strikingly.data.logappender

import java.io.Serializable

import com.amazonaws.services.sqs.{AmazonSQS, AmazonSQSClientBuilder}
import org.apache.logging.log4j.core.{Filter, Layout, LogEvent}
import org.apache.logging.log4j.core.appender.AbstractAppender
import org.apache.logging.log4j.core.config.plugins.Plugin
import org.apache.logging.log4j.core.config.plugins.PluginAttribute
import org.apache.logging.log4j.core.config.plugins.PluginElement
import org.apache.logging.log4j.core.config.plugins.PluginFactory
import org.apache.logging.log4j.status.StatusLogger

import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.util.Try

object SqsAppender {
  lazy val sqsClient: AmazonSQS = {
    val region = {
      val _region = System.getProperty("AWS_REGION")
      if (_region == null) "cn-north-1"
      else _region
    }
    AmazonSQSClientBuilder.standard().withRegion(region).build()
  }
  var sqsUrl: Option[String] = None

  def initSqsUrl(queueName: String): Unit = {
    if (sqsUrl.isEmpty) {
      val url = Try[String] {
        SqsAppender.sqsClient.getQueueUrl(queueName).getQueueUrl
      }
      sqsUrl = Option(url.getOrElse(null))
    }
  }

  @PluginFactory
  def createAppender(@PluginAttribute("name") name: String,
                     @PluginAttribute("QueueName") queueName: String,
                     @PluginElement("Filter") filter: Filter,
                     @PluginElement("Layout") layout: Layout[_ <: Serializable],
                     @PluginAttribute("ignoreExceptions") ignoreExceptions: Boolean): SqsAppender = {
    if (queueName == null) {
      StatusLogger.getLogger.error("no queue name provided")
      return null
    }

    StatusLogger.getLogger.info(s"Get SQS queue name: $queueName")
    initSqsUrl(queueName)

    new SqsAppender(name, filter, layout, ignoreExceptions)
  }
}

@Plugin(name = "SqsAppender", category = "Core", elementType = "appender", printObject = true)
class SqsAppender(name: String, filter: Filter, layout: Layout[_ <: Serializable],
                  ignoreExceptions: Boolean) extends AbstractAppender(name, filter, layout, ignoreExceptions) {
  override def append(event: LogEvent): Unit = {
    import com.amazonaws.services.sqs.model.SendMessageRequest

    SqsAppender.sqsUrl.foreach { url =>
      val msg = getLayout.toSerializable(event).toString
      Future { SqsAppender.sqsClient.sendMessage(new SendMessageRequest(url, msg)) }
    }
  }
}
